src/main/java/com/minecolonies/coremod/entity/ai/citizen/herders/AbstractEntityAIHerder.java

```diff
@@ -0,0 +1,534 @@
+     *
+     * @return The next {@link AIState} the herder should switch to, after executing this method.
+     */
+    public AIState decideWhatToDo()
+    {
+        setDelay(DELAY_FOURTY);
+
+        final List<T> animals = new ArrayList<>(searchForAnimals());
+
+        if (animals.isEmpty())
+        {
+            setDelay(DELAY_ONE_HUNDRED);
+            return HERDER_DECIDE;
+        }
+
+        worker.setLatestStatus(new TextComponentTranslation("com.minecolonies.coremod.status.herder.deciding"));
+
+        final int numOfBreedableAnimals = animals.stream().filter(animal -> animal.getGrowingAge() == 0).toArray().length;
+
+        final boolean hasBreedingItem = worker.hasItemInInventory(getBreedingItem(), 0);
+
+        if (!searchForItemsInArea().isEmpty())
+        {
+            return HERDER_PICKUP;
+        }
+        else if (maxAnimals())
+        {
+            return HERDER_BUTCHER;
+        }
+        else if (numOfBreedableAnimals >= NUM_OF_ANIMALS_TO_BREED && hasBreedingItem)
+        {
+            return HERDER_BREED;
+        }
+        return START_WORKING;
+    }
+
+    /**
+     * Redirects the herder to their building.
+     *
+     * @return The next {@link AIState}.
+     */
+    private AIState startWorkingAtOwnBuilding()
+    {
+        worker.setLatestStatus(new TextComponentTranslation("com.minecolonies.coremod.status.herder.goingToHut"));
+        if (walkToBuilding())
+        {
+            return getState();
+        }
+        return PREPARING;
+    }
+
+    /**
+     * Prepares the herder for herding
+     *
+     * @return The next {@link AIState}.
+     */
+    private AIState prepareForHerding()
+    {
+        toolsNeeded.add(ToolType.AXE);
+        itemsNeeded.add(new ItemStack(getBreedingItem(), 2));
+
+        for (final ToolType tool : toolsNeeded)
+        {
+            if (checkForToolOrWeapon(tool))
+            {
+                return getState();
+            }
+        }
+
+        for (final ItemStack item : itemsNeeded)
+        {
+            checkOrRequestItemsAsynch(false, item);
+        }
+
+        return HERDER_DECIDE;
+    }
+
+    /**
+     * Butcher some animals (Preferably Adults) that the herder looks after.
+     *
+     * @return The next {@link AIState}.
+     */
+    private AIState butcherAnimals()
+    {
+        setDelay(DELAY_TWENTY);
+
+        if (!maxAnimals())
+        {
+            return HERDER_DECIDE;
+        }
+
+        if (!equipTool(ToolType.AXE))
+        {
+            return START_WORKING;
+        }
+
+        EntityAnimal animal = searchForAnimals().stream().filter(animalToButcher -> !animalToButcher.isChild()).findFirst().orElse(null);
+
+        if (animal == null)
+        {
+            animal = searchForAnimals().stream().findFirst().orElse(null);
+        }
+
+        butcherAnimal(animal);
+
+        if (animal != null && !animal.isEntityAlive())
+        {
+            incrementActionsDone();
```

Comment:
will this pick the same animal again or will he hit random animals till the first few drop dead?

Reply:
Same animal each time until that animal dies